### DISCUSSION TITLE
Dependencies Injection on IDocumentSessionListener

### DISCUSSION THREAD

**Discussion by eliecer2323** [JasperFx/marten#1878] [Q&A]
Hello everyone, I'm trying to integrate an interceptor of CRUD operation to an specific table for doing another operations, so I found the IDocumentSessionListener interface for doing it

Here is my custom listener:
```
public class LedgerListener : IDocumentSessionListener
    {
        private readonly ICSVService _csvSvc;

        public LedgerListener(ICSVService csvSvc)
        {
            _csvSvc = csvSvc;
        }

        public async Task BeforeSaveChangesAsync(IDocumentSession session, CancellationToken token)
        {
            IUnitOfWork pendingChanges = session.PendingChanges;

            IEnumerable<Ledger> pendingLedgerInserts = pendingChanges.InsertsFor<Ledger>();

            foreach (Ledger record in pendingLedgerInserts)
            {
                _csvSvc.CreateSheet(record);
            }
        }
}
```

My service:
```
public class CSVService : ICSVService
{
        private readonly IDocumentStore _documentStore;

        public CSVService(IDocumentStore documentStore)
        {
            _documentStore = documentStore;
        }

        public void CreateSheet(Ledger contract)
        {
            using (var session = _documentStore.LightweightSession())
            {
                Console.WriteLine("Some proccess ...");
            }
        }
}
```

And here is the listener additions in the Startup.cs:
```
        ...previous code...
        services.AddTransient<ICSVService, CSVService>();
        services.AddScoped<IDocumentStore>(provider =>
                DocumentStore.For(_ =>
                {
                    _.AutoCreateSchemaObjects = AutoCreate.CreateOrUpdate;
                    _.Connection(cn);
                    _.Listeners.Add(new LedgerListener());
                }));
        ...next code...
```

But my current trouble is that I don't find any way to do the dependencies injection for the ICSVService and its dependency IDocumentStore, is there a way to do the CDI in the `_.Listeners.Add(new LedgerListener());` code line? I can't simply pass the instances manually because there could be too many dependencies with their own dependencies

**Comment by jeremydmiller** [comment]
@eliecer2323 You can add additional document session listeners to an `IDocumentSession` **after** it's created. So one way to do what I think you need to do here is change your IoC registrations for `IDocumentSession` to add the necessary listeners inside a Lambda service registration. You can utilize the DI container right then and there to pull out whatever you need.

If you're in v4, you can use a custom factory to build document sessions I can dig up for you if you want. We're **still** updating docs for v4 and we're behind.

**Comment by eliecer2323** [comment]
Hi @jeremydmiller thanks for your support I appreciate it.

do you suggest to use the `IDocumentSession` instead of the `IDocumentStore `, but then I'd need to add the listeners on each place I instantiate a IDocumentSession and my objective is to subscribe a process to CRUD operations to that specific table independently on who does that without the need to have knowledge of this. I didn't get your thought about pulling out the dependencies from the DI container, I have very limited knowledges about .Net, could you dig in it please and maybe give me a code example please?

**Reply by amferguson** [reply]
Late to the party, but I had this same question a few days ago, and this discussion pointed me in the right direction (thanks!). This example is specifically for Autofac, but @jeremydmiller's suggestion to do it as a lambda registration ought to translate to any other DI system that supports it. I did something (basically) like this:

```
builder.Register(r =>
   {
       var session = r.Resolve<IDocumentStore>().DirtyTrackedSession();
       session.Listeners.Add(r.Resolve<ApplicationStoreListener>()); //ApplicationStoreListener requires other services 
       return session;
   }).As<IDocumentSession>().InstancePerLifetimeScope();
```

where `builder` is the Autofac `ContainerBuilder` and `ApplicationStoreListener` is an implementation of `IDocumentSessionListener` that requires other services that are registered with the container. Then, wherever we want this type of session, we just inject an `IDocumentSession` and get one with the listener added to its list.

**Reply by jeremydmiller** [reply]
There's an overload of `AddMarten()` that lets you supply a `Func<IServiceProvider, StoreOptions>` model to configure Marten. That would let you set up whatever you want here. That or you can supply your own `ISessionBuilder` if the listener is per session.
